File: wnode-sdk-go/mesh/integrity.go
package mesh

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
)

type IntegrityProtectedMessage struct {
	MessageID          string
	SenderNodeID       string
	PayloadHash        string
	Payload            interface{}
	IntegrityProof     string
	SenderCapabilities []string
}

type MeshIntegrityValidator struct {
	authRegistry *MeshAuthRegistry
	macSecret    string
}

func NewMeshIntegrityValidator(auth *MeshAuthRegistry) *MeshIntegrityValidator {
	return &MeshIntegrityValidator{
		authRegistry: auth,
		macSecret:    "mesh-phase1.6-mac-secret",
	}
}

func (v *MeshIntegrityValidator) ValidateMessage(msg IntegrityProtectedMessage) error {
	if !v.authRegistry.IsTrusted(msg.SenderNodeID) {
		return errors.New("INTEGRITY_REJECTED: Sender not trusted")
	}

	b, _ := json.Marshal(msg.Payload)
	hash := sha256.Sum256(b)
	hashHex := hex.EncodeToString(hash[:])

	if hashHex != msg.PayloadHash {
		v.authRegistry.MarkSuspicious(msg.SenderNodeID)
		return errors.New("INTEGRITY_REJECTED: Payload hash mismatch")
	}

	expectedProof := v.ComputeIntegrityProof(msg.MessageID, msg.PayloadHash, msg.SenderNodeID)
	if msg.IntegrityProof != expectedProof {
		v.authRegistry.MarkSuspicious(msg.SenderNodeID)
		return errors.New("INTEGRITY_REJECTED: Invalid integrity proof")
	}

	return nil
}

func (v *MeshIntegrityValidator) ComputeIntegrityProof(messageID, payloadHash, senderNodeID string) string {
	data := messageID + ":" + payloadHash + ":" + senderNodeID
	mac := hmac.New(sha256.New, []byte(v.macSecret))
	mac.Write([]byte(data))
	return hex.EncodeToString(mac.Sum(nil))
}

File: packages/wnode-sdk-ts/src/mesh/integrity.ts
import * as crypto from 'crypto';
import { MeshAuthRegistry } from './auth';
import { GossipMessage } from './types';
import { WnodeDeterminismError } from '../errors';

export interface IntegrityProtectedMessage extends GossipMessage {
  integrityProof: string;
  senderCapabilities: string[];
}

export class MeshIntegrityValidator {
  // Config-derived MAC secret for integrity checks
  private readonly macSecret = 'mesh-phase1.6-mac-secret';

  constructor(private readonly authRegistry: MeshAuthRegistry) {}

  public validateMessage(message: IntegrityProtectedMessage): void {
    // 1. Validate sender is trusted
    if (!this.authRegistry.isTrusted(message.senderNodeId)) {
      throw new WnodeDeterminismError('INTEGRITY_REJECTED', {
        reason: 'Sender node is not trusted or is suspicious',
        senderNodeId: message.senderNodeId
      });
    }

    // 2. Validate payload hash matches payload
    const computedHash = crypto.createHash('sha256').update(JSON.stringify(message.payload)).digest('hex');
    if (computedHash !== message.payloadHash) {
      this.authRegistry.markSuspicious(message.senderNodeId);
      throw new WnodeDeterminismError('INTEGRITY_REJECTED', {
        reason: 'Payload hash mismatch - potential tampering',
        senderNodeId: message.senderNodeId
      });
    }

    // 3. Validate integrity proof (MAC)
    const expectedProof = this.computeIntegrityProof(message.messageId, message.payloadHash, message.senderNodeId);
    if (message.integrityProof !== expectedProof) {
      this.authRegistry.markSuspicious(message.senderNodeId);
      throw new WnodeDeterminismError('INTEGRITY_REJECTED', {
        reason: 'Invalid integrity proof - message integrity compromised',
        senderNodeId: message.senderNodeId
      });
    }
  }

  public computeIntegrityProof(messageId: string, payloadHash: string, senderNodeId: string): string {
    const data = `${messageId}:${payloadHash}:${senderNodeId}`;
    return crypto.createHmac('sha256', this.macSecret).update(data).digest('hex');
  }

  public signMessage(msg: GossipMessage, senderCapabilities: string[]): IntegrityProtectedMessage {
    const integrityProof = this.computeIntegrityProof(msg.messageId, msg.payloadHash, msg.senderNodeId);
    return {
      ...msg,
      integrityProof,
      senderCapabilities
    };
  }
}
